## Elias Villaseñor Flores
## 1735689
## Estadística Aplicada
## Grupo 51

In [58]:
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt
import statsmodels.formula.api as smf
%matplotlib inline

In [4]:
datos=pd.read_excel('modelolineal.xlsx')

In [6]:
datos['X2']= datos['X']**2
datos['X3']= datos['X']**3
datos['X4']= datos['X']**4
datos

observación          Y   X   X2     X3      X4
0             1  19.246957  23  529  12167  279841
1             2  16.878501  19  361   6859  130321
2             3  20.187237  27  729  19683  531441
3             4  12.176789  13  169   2197   28561
4             5  21.420112  21  441   9261  194481
5             6  24.436799  22  484  10648  234256
6             7  14.507928  19  361   6859  130321
7             8  16.013496  14  196   2744   38416
8             9  12.043514   3    9     27      81
9            10   6.808838   1    1      1       1
10           11  20.473608  19  361   6859  130321
11           12  19.688717  18  324   5832  104976
12           13  20.405782  12  144   1728   20736
13           14  24.959259  16  256   4096   65536
14           15  23.913532  29  841  24389  707281
15           16  22.719407  23  529  12167  279841
16           17  18.680587  14  196   2744   38416
17           18  17.192583  20  400   8000  160000
18           19  28.422786  27  729  19683  531441
19           20  24.757379  28  784  21952  614656
20           21   9.544102   0    0      0       0
21           22  24.230330  20  400   8000  160000
22           23  26.417826  24  576  13824  331776
23           24  23.066723  23  529  12167  279841
24           25  13.145005   8   64    512    4096
25           26  24.385156  24  576  13824  331776
26           27  12.934102  10  100   1000   10000
27           28  14.002469   4   16     64     256
28           29  26.599047  29  841  24389  707281
29           30  11.910387  10  100   1000   10000

## Regresión 1

### a)Divide los datosen: conjunto de entrenamientoalas observaciones 1-20 (train_set) y conjunto de prueba alas observaciones 21-30 (test_set).

In [8]:
train_set = datos.iloc[:20]
test_set = datos.iloc[20:]

### b)Estima el siguiente modelolineal con el conjunto de entrenamiento𝑦=𝑏1𝑥+𝜖Ymuestra el resultado de la regresión(print_summary())

In [10]:
reg1 = smf.ols(formula='Y~X', data=train_set)
results1 = reg1.fit()
print(results1.summary())

                            OLS Regression Results                            
Dep. Variable:                      Y   R-squared:                       0.615
Model:                            OLS   Adj. R-squared:                  0.594
Method:                 Least Squares   F-statistic:                     28.80
Date:                Mon, 12 Sep 2022   Prob (F-statistic):           4.23e-05
Time:                        15:40:24   Log-Likelihood:                -51.332
No. Observations:                  20   AIC:                             106.7
Df Residuals:                      18   BIC:                             108.7
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      9.2336      2.008      4.598      0.0

In [12]:
test_set['y_gor1'] = results1.predict(test_set)
train_set['y_gor1']= results1.predict(train_set)

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_5840\1145136023.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  test_set['y_gor1'] = results1.predict(test_set)
C:\Users\Lenovo\AppData\Local\Temp\ipykernel_5840\1145136023.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  train_set['y_gor1']= results1.predict(train_set)


In [16]:
test_set[['Y','y_gor1']]

Y     y_gor1
20   9.544102   9.233587
21  24.230330  20.117395
22  26.417826  22.294157
23  23.066723  21.749966
24  13.145005  13.587110
25  24.385156  22.294157
26  12.934102  14.675491
27  14.002469  11.410349
28  26.599047  25.015109
29  11.910387  14.675491

### c)Con la regresión anterior pronosticael valor de𝑦̂ tanto en el conjunto de prueba como  en  el  conjunto  de  entrenamiento. Del  MAPE,  MAD  y  MSD elige  dos  y obténsu  valorpara  el  conjunto de  pruebay  el  conjunto  de entrenamiento.

In [64]:
from sklearn.metrics import mean_absolute_error as MAD 
from sklearn.metrics import mean_absolute_percentage_error as MAPE

In [65]:
MAD_E1 = MAD(train_set['Y'],train_set['y_gor1'])
MAD_P1 = MAD(test_set['Y'],test_set['y_gor1'])

MAPE_E1 = MAPE(train_set['Y'],train_set['y_gor1'])
MAPE_P1 = MAPE(test_set['Y'],test_set['y_gor1'])

In [24]:
round(MAD_E1,3), round(MAD_P1,3)

(2.597, 2.108)

In [67]:
round(MAPE_E1, 3), round(MAPE_P1, 3)

(0.151, 0.115)

### d)Estima el siguiente modelolineal con el conjunto de entrenamiento𝑦=𝑏1𝑥+𝑏2𝑥2+𝜖

In [28]:
reg2 = smf.ols(formula='Y~X+X2', data=train_set)
results2 = reg2.fit()


In [39]:
test_set['y_gor2'] = results2.predict(test_set)
train_set['y_gor2']= results2.predict(train_set)

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_5840\3572963804.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  test_set['y_gor2'] = results2.predict(test_set)
C:\Users\Lenovo\AppData\Local\Temp\ipykernel_5840\3572963804.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  train_set['y_gor2']= results2.predict(train_set)


In [68]:
MAD_E2 = MAD(train_set['Y'],train_set['y_gor2'])
MAD_P2 = MAD(test_set['Y'],test_set['y_gor2'])

MAPE_E2 = MAPE(train_set['Y'],train_set['y_gor2'])
MAPE_P2 = MAPE(test_set['Y'],test_set['y_gor2'])

In [70]:
round(MAD_E2,3), round(MAD_P2,3)

(2.554, 2.334)

In [71]:
round(MAPE_E2, 3), round(MAPE_P2, 3)

(0.147, 0.133)

## e)Estima el siguiente modelolineal con el conjunto de entrenamiento 𝑦=𝑏1𝑥+𝑏2𝑥2+𝑏3𝑥3+𝜖

In [44]:
reg3 = smf.ols(formula='Y~X+X2+X3', data=train_set)
results3 = reg3.fit()

In [46]:
test_set['y_gor3'] = results3.predict(test_set)
train_set['y_gor3']= results3.predict(train_set)

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_5840\523141156.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  test_set['y_gor3'] = results3.predict(test_set)
C:\Users\Lenovo\AppData\Local\Temp\ipykernel_5840\523141156.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  train_set['y_gor3']= results3.predict(train_set)


In [73]:
MAD_E3 = MAD(train_set['Y'],train_set['y_gor3'])
MAD_P3 = MAD(test_set['Y'],test_set['y_gor3'])

MAPE_E3 = MAPE(train_set['Y'],train_set['y_gor3'])
MAPE_P3 = MAPE(test_set['Y'],test_set['y_gor3'])

In [50]:
round(MAD_E3,3), round(MAD_P3,3)

(2.531, 3.013)

In [74]:
round(MAPE_E3, 3), round(MAPE_P3, 3)

(0.141, 0.186)

### f)Estima el siguiente modelolineal con el conjunto de entrenamiento𝑦=𝑏1𝑥+𝑏2𝑥2+𝑏3𝑥3+𝑏4𝑥4+𝜖

In [51]:
reg4 = smf.ols(formula='Y~X+X2+X3+X4', data=train_set)
results4 = reg4.fit()

In [53]:
test_set['y_gor4'] = results4.predict(test_set)
train_set['y_gor4']= results4.predict(train_set)

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_5840\4285325944.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  test_set['y_gor4'] = results4.predict(test_set)
C:\Users\Lenovo\AppData\Local\Temp\ipykernel_5840\4285325944.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  train_set['y_gor4']= results4.predict(train_set)


In [55]:
MAD_E4 = MAD(train_set['Y'],train_set['y_gor4'])
MAD_P4 = MAD(test_set['Y'],test_set['y_gor4'])

MAPE_E4 = MAPE(train_set['Y'],train_set['y_gor4'])
MAPE_P4 = MAPE(test_set['Y'],test_set['y_gor4'])

In [59]:
round(MAD_E4,3), round(MAD_P4,3)

(2.34, 3.507)

In [75]:
round(MAPE_E3, 3), round(MAPE_P3, 3)

(0.141, 0.186)

### g)Realiza un cuadro como el siguientey llénalocon las pruebas que hayas elegido, por ejemplo,si elegiste MAPE y MSD:

|                 MAD                     |                               |                        |
|:-----------------------------------:    |:-------------------------:    |:------------------:    |
|              Regresión                  | Conjunto de entrenamiento     | Conjunto de prueba     |
| y = b1x + e                             |           2.597               |        2.108           |
| y = b1x + b2x^2 + e                     |           2.554               |        2.334           |
| y = b1x + b2x^2 + b3x^3 + e             |           2.531               |        3.013           |
| y = b1x + b2x^2 + b3x^3 + b4x^4 + e     |            2.34               |        3.507           |

|                 MAPE                    |                               |                        |
|:-----------------------------------:    |:-------------------------:    |:------------------:    |
|              Regresión                  | Conjunto de entrenamiento     | Conjunto de prueba     |
| y = b1x + e                             |           0.151               |        0.115           |
| y = b1x + b2x^2 + e                     |           0.147               |        0.133           |
| y = b1x + b2x^2 + b3x^3 + e             |           0.141               |        0.186           |
| y = b1x + b2x^2 + b3x^3 + b4x^4 + e     |           0.141               |        0.186           |